File: WoofIOS/SwiftPackages/DesignSystem/Sources/DesignSystem/Atomic/Organisms/ActionBar/OrgTypeBarWithAction.swift
import Resources
import SwiftUI

public struct OrgTypeBarWithAction: View {
    private let text: String
    private let placeholder: String
    private let onTextChange: (String) -> Void
    private let onSubmit: () -> Void
    private let actionIcon: ImageAsset
    private let actionContentDescription: String
    private let actionColorRole: IconColorRole
    private let onActionTap: () -> Void
    private let isSubmitEnabled: Bool

    public init(
        text: String,
        placeholder: String,
        onTextChange: @escaping (String) -> Void,
        onSubmit: @escaping () -> Void,
        actionIcon: ImageAsset,
        actionContentDescription: String,
        actionColorRole: IconColorRole,
        onActionTap: @escaping () -> Void,
        isSubmitEnabled: Bool
    ) {
        self.text = text
        self.placeholder = placeholder
        self.onTextChange = onTextChange
        self.onSubmit = onSubmit
        self.actionIcon = actionIcon
        self.actionContentDescription = actionContentDescription
        self.actionColorRole = actionColorRole
        self.onActionTap = onActionTap
        self.isSubmitEnabled = isSubmitEnabled
    }

    public var body: some View {
        HStack(spacing: SpacingRoles.Component.compact.rawValue) {
            MolTypeBar(
                text: text,
                placeholder: placeholder,
                onTextChange: onTextChange,
                onSubmit: onSubmit,
                isSubmitEnabled: isSubmitEnabled
            )
            AtomIconButton(
                icon: actionIcon,
                contentDescription: actionContentDescription,
                colorRole: actionColorRole,
                backgroundRole: .scrimContainer,
                onTap: onActionTap
            )
        }
        .padding(.horizontal, PaddingRoles.Element.relaxed.rawValue)
        .padding(.vertical, PaddingRoles.Element.regular.rawValue)
    }
}

File: WoofAndroid/design-system/src/main/kotlin/com/perrystreet/woof/designsystem/atomic/organisms/actionbar/OrgTypeBarWithAction.kt
package com.perrystreet.woof.designsystem.atomic.organisms.actionbar

import androidx.annotation.DrawableRes
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.PaddingRoles
import com.perrystreet.woof.designsystem.atomic._tokens.spacing.SpacingRoles
import com.perrystreet.woof.designsystem.atomic.atoms.button.AtomIconButton
import com.perrystreet.woof.designsystem.atomic.atoms.button.roles.ButtonBackgroundRole
import com.perrystreet.woof.designsystem.atomic.atoms.icon.roles.IconColorRole
import com.perrystreet.woof.designsystem.atomic.molecules.typebar.MolTypeBar

@Composable
fun OrgTypeBarWithAction(
    text: String,
    placeholder: String,
    onTextChange: (String) -> Unit,
    onSubmit: () -> Unit,
    @DrawableRes actionIconRes: Int,
    actionContentDescription: String,
    actionColorRole: IconColorRole,
    onActionTap: () -> Unit,
    isSubmitEnabled: Boolean = text.isNotBlank(),
) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(
                horizontal = PaddingRoles.Element.Relaxed.dp,
                vertical = PaddingRoles.Element.Regular.dp,
            ),
        horizontalArrangement = Arrangement.spacedBy(SpacingRoles.Component.Compact.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        MolTypeBar(
            text = text,
            placeholder = placeholder,
            onTextChange = onTextChange,
            onSubmit = onSubmit,
            isSubmitEnabled = isSubmitEnabled,
            modifier = Modifier.weight(1f),
        )
        AtomIconButton(
            iconRes = actionIconRes,
            contentDescription = actionContentDescription,
            colorRole = actionColorRole,
            backgroundRole = ButtonBackgroundRole.ScrimContainer,
            onTap = onActionTap,
        )
    }
}
